File: assets/scripts/ocr.py

import sys
import asyncio
from pathlib import Path

# Try to import winsdk
try:
    from winsdk.windows.media.ocr import OcrEngine
    from winsdk.windows.graphics.imaging import BitmapDecoder, SoftwareBitmap
    from winsdk.windows.storage import StorageFile, FileAccessMode
    from winsdk.windows.storage.streams import IRandomAccessStream
    from winsdk.windows.globalization import Language
except ImportError:
    print("Error: winsdk not installed. Please pip install winsdk")
    sys.exit(1)

async def run_ocr(image_path):
    try:
        # Load File
        file = await StorageFile.get_file_from_path_async(str(Path(image_path).resolve()))
        stream = await file.open_async(FileAccessMode.READ)
        
        # Decode
        decoder = await BitmapDecoder.create_async(stream)
        # OcrEngine requires input to be less than MaxImageDimension?
        # Usually fine for 1080p.
        
        soft_bmp = await decoder.get_software_bitmap_async()
        
        # Create Engine
        # Try English first
        lang = Language("en-US")
        engine = None
        if OcrEngine.is_language_supported(lang):
             engine = OcrEngine.try_create_from_language(lang)
        
        if not engine:
             engine = OcrEngine.try_create_from_user_profile_languages()
             
        if not engine:
            print("Error: Could not create OCR Engine.")
            return

        # Recognize
        result = await engine.recognize_async(soft_bmp)
        
        # Print
        if result and result.lines:
            for line in result.lines:
                print(line.text)
        
    except Exception as e:
        print(f"Error: {e}")

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python ocr.py <image_path>")
        sys.exit(1)
        
    path = sys.argv[1]
    asyncio.run(run_ocr(path))

File: assets/scripts/ocr.cs
using System;
using System.IO;
using System.Threading;
using Windows.Graphics.Imaging;
using Windows.Media.Ocr;
using Windows.Storage;
using Windows.Storage.Streams;
using Windows.Foundation;

namespace AutoLauncher.OCR
{
    class Program
    {
        static void Main(string[] args)
        {
            if (args.Length == 0)
            {
                Console.WriteLine("Usage: ocr.exe <image_path>");
                return;
            }

            string imagePath = args[0];
            try
            {
                // Initialize OCR Engine
                // Try to force English first
                OcrEngine ocrEngine = null;
                var lang = new Windows.Globalization.Language("en-US");
                if (OcrEngine.IsLanguageSupported(lang))
                {
                    ocrEngine = OcrEngine.TryCreateFromLanguage(lang);
                }
                
                if (ocrEngine == null)
                {
                    ocrEngine = OcrEngine.TryCreateFromUserProfileLanguages();
                }

                if (ocrEngine == null)
                {
                    Console.WriteLine("Error: Could not create OCR Engine (Language not supported?)");
                    return;
                }
                Console.WriteLine("DEBUG: Engine Created. Language: " + ocrEngine.RecognizerLanguage.DisplayName);

                // Load Image
                var fileOp = StorageFile.GetFileFromPathAsync(Path.GetFullPath(imagePath));
                var file = Await(fileOp);

                var streamOp = file.OpenAsync(FileAccessMode.Read);
                using (var stream = Await(streamOp))
                {
                    var decoderOp = BitmapDecoder.CreateAsync(stream);
                    var decoder = Await(decoderOp);

                    Console.WriteLine("DEBUG: Image Decoded " + decoder.PixelWidth + "x" + decoder.PixelHeight);

                    var bitmapOp = decoder.GetSoftwareBitmapAsync();
                    var softwareBitmap = Await(bitmapOp);

                    var ocrOp = ocrEngine.RecognizeAsync(softwareBitmap);
                    var ocrResult = Await(ocrOp);

                    Console.WriteLine("DEBUG: Found " + ocrResult.Lines.Count + " lines.");

                    foreach (var line in ocrResult.Lines)
                    {
                        Console.WriteLine(line.Text);
                    }
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine("Error: " + ex.Message);
            }
        }

        // Helper to await WinRT async operations synchronously
        static T Await<T>(IAsyncOperation<T> op)
        {
            // Simple polling wait
            while (op.Status == AsyncStatus.Started)
            {
                Thread.Sleep(10);
            }

            if (op.Status == AsyncStatus.Completed)
            {
                return op.GetResults();
            }
            
            if (op.Status == AsyncStatus.Error)
            {
                throw new Exception("Async Operation Failed: " + op.ErrorCode.Message);
            }

            throw new Exception("Async Operation Canceled or Failed");
        }
    }
}
